[ao] Standardize PerChannelDetector Output to be single level

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81647
* #81589
* #81523
* #81588
* #81587
* #81586
* __->__ #81585

Summary: Currently, the PerChannelDetector has a multi-layered output.

Example:
```
{'backend': 'qnnpack',
 'per_channel_status': {'block1.linear': {'per_channel_supported': True,
                                          'per_channel_used': False},
                        'block2.linear': {'per_channel_supported': True,
                                          'per_channel_used': False}}}
```

The issue with this is that when it comes to future features such as
visualizations where we need to go through this dictionary, it can be
hard because of the variable number of layers.

This changes the output format of the PerChannelDetector to have a
standard format.

Ex.)
```
{'block1.linear': {'backend': 'qnnpack',
                   'per_channel_supported': True,
                   'per_channel_used': False},
 'block2.linear': {'backend': 'qnnpack',
                   'per_channel_supported': True,
                   'per_channel_used': False}}
```

Test Plan: python test/test_quantization.py TestFxModelReportDetector

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Standardize PerChannelDetector Output to be single level

Summary: Currently, the PerChannelDetector has a multi-layered output.

Example:
```
{'backend': 'qnnpack',
 'per_channel_status': {'block1.linear': {'per_channel_supported': True,
                                          'per_channel_used': False},
                        'block2.linear': {'per_channel_supported': True,
                                          'per_channel_used': False}}}
```

The issue with this is that when it comes to future features such as
visualizations where we need to go through this dictionary, it can be
hard because of the variable number of layers.

This changes the output format of the PerChannelDetector to have a
standard format.

Ex.)
```
module_fqn
|
--- feature name: val
```

Test Plan: python test/test_quantization.py TestFxModelReportDetector

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Standardize PerChannelDetector Output to be single level"


Summary: Currently, the PerChannelDetector has a multi-layered output.

Example:
```
{'backend': 'qnnpack',
 'per_channel_status': {'block1.linear': {'per_channel_supported': True,
                                          'per_channel_used': False},
                        'block2.linear': {'per_channel_supported': True,
                                          'per_channel_used': False}}}
```

The issue with this is that when it comes to future features such as
visualizations where we need to go through this dictionary, it can be
hard because of the variable number of layers.

This changes the output format of the PerChannelDetector to have a
standard format.

Ex.)
```
module_fqn
|
--- feature name: val
```

Test Plan: python test/test_quantization.py TestFxModelReportDetector

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Standardize PerChannelDetector Output to be single level"


Summary: Currently, the PerChannelDetector has a multi-layered output.

Example:
```
{'backend': 'qnnpack',
 'per_channel_status': {'block1.linear': {'per_channel_supported': True,
                                          'per_channel_used': False},
                        'block2.linear': {'per_channel_supported': True,
                                          'per_channel_used': False}}}
```

The issue with this is that when it comes to future features such as
visualizations where we need to go through this dictionary, it can be
hard because of the variable number of layers.

This changes the output format of the PerChannelDetector to have a
standard format.

Ex.)
```
module_fqn
|
--- feature name: val
```

Test Plan: python test/test_quantization.py TestFxModelReportDetector

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -189,9 +192,10 @@ def _detect_per_channel_helper(self, model: nn.Module, per_channel_info: Dict):
+                per_channel_info[fqn] = {
+                    self.PER_CHAN_SUPPORTED_KEY: per_channel_supported,
+                    self.PER_CHAN_USED_KEY: per_channel_used,
+                    self.BACKEND_KEY: self.backend_chosen
```
nit: not sure if there's a more elegant way to do this since its more of a model level piece of info than a module level...

vspenubarthi replied:
i initially had it at a higher level since it was a model level piece and not module specific, but that was the issue in terms of consistent "levels" for features across all the different detectors, which is why i just decided to include it at this level. It makes the Visualizer much easier

HDCharles reviewed (approved):
good subject to better example in summary

vspenubarthi committed:
Update on "[ao] Standardize PerChannelDetector Output to be single level"


Summary: Currently, the PerChannelDetector has a multi-layered output.

Example:
```
{'backend': 'qnnpack',
 'per_channel_status': {'block1.linear': {'per_channel_supported': True,
                                          'per_channel_used': False},
                        'block2.linear': {'per_channel_supported': True,
                                          'per_channel_used': False}}}
```

The issue with this is that when it comes to future features such as
visualizations where we need to go through this dictionary, it can be
hard because of the variable number of layers.

This changes the output format of the PerChannelDetector to have a
standard format.

Ex.)
```
{'block1.linear': {'backend': 'qnnpack',
                   'per_channel_supported': True,
                   'per_channel_used': False},
 'block2.linear': {'backend': 'qnnpack',
                   'per_channel_supported': True,
                   'per_channel_used': False}}
```

Test Plan: python test/test_quantization.py TestFxModelReportDetector

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
